File: src/ios/admob/AdMobInterstitial.m
//
//  AdMob ANE Extension
//  iOS Native Extension
//
//  AdMobInterstitial.m
//  AdMobAne
//
//  Copyright (c) 2011-2015 lancelot Inc. All rights reserved.
//

#import "FlashRuntimeExtensions.h"
#import "GADInterstitial.h"
#import "GADInterstitialDelegate.h"
#import "AdMobAne.h"
#import "AdMobInterstitial.h"
#import "AdMobEvents.h"

@implementation AdMobInterstitial

// AdMob Banner Constants
const NSString *INTER_ID = @"InterstitialAd";

// AdMob Interstitial Propriety
@synthesize context;
@synthesize adMobRequest;
@synthesize adMobId;

// =================================================================================================
//	Admob Delegate Events
// =================================================================================================

/**
 * Interstitial Did Receive Ad Event Listener
 *
 */
- (void)interstitialDidReceiveAd:(GADInterstitial *)interstitial
{
    Log(@"Event interstitialDidReceiveAd");
    
    // Update Interstitial load state
    mIsInterstitialLoaded = YES;
    // Dispatch the Event
    FREDispatchStatusEventAsync(self.context, (uint8_t*)INTERSTITIAL_LOADED, (uint8_t*)[INTER_ID UTF8String]);
}

/**
 * Did Fail To Receive Interstitial With Error Event Listener
 *
 */
- (void)interstitial:(GADInterstitial *)interstitial didFailToReceiveAdWithError:(GADRequestError *)error;
{
    Log(@"Event didFailToReceiveAdWithError");
    Log([NSString stringWithFormat:@"Error Details: %@",error]);
    
    // Dispatch the Event
    FREDispatchStatusEventAsync(self.context, (uint8_t*)INTERSTITIAL_FAILED_TO_LOAD, (uint8_t*)[INTER_ID UTF8String]);
    // Clear the Interstitial from memory
    [self clearInterstitial];
}

/**
 * Interstitial Will Present Screen Event Listener
 *
 */
- (void)interstitialWillPresentScreen:(GADInterstitial *)interstitial;
{
    Log(@"Event interstitialWillPresentScreen");
    
    // Dispatch the Event
    FREDispatchStatusEventAsync(self.context, (uint8_t*)INTERSTITIAL_AD_OPENED, (uint8_t*)[INTER_ID UTF8String]);
}

/**
 * Interstitial Will Dismiss Screen Event Listener
 *
 */
- (void)interstitialWillDismissScreen:(GADInterstitial *)interstitial;
{
    Log(@"Event interstitialWillDismissScreen");
    /*
    // Dispatch the Event
    FREDispatchStatusEventAsync(self.context, (uint8_t*)INTERSTITIAL_AD_CLOSED, (uint8_t*)[INTER_ID UTF8String]);
    */
}

/**
 * Interstitial Did Dismiss Screen Event Listener
 *
 */
- (void)interstitialDidDismissScreen:(GADInterstitial *)interstitial;
{
    Log(@"Event interstitialDidDismissScreen");
    
    // Dispatch the Event
    FREDispatchStatusEventAsync(self.context, (uint8_t*)INTERSTITIAL_AD_CLOSED, (uint8_t*)[INTER_ID UTF8String]);
    // Clear the Interstitial from memory
    [self clearInterstitial];
}

/**
 * Interstitial Will Leave Application Event Listener
 *
 */
- (void)interstitialWillLeaveApplication:(GADInterstitial *)interstitial;
{
    Log(@"Event interstitialWillLeaveApplication");
    
    // Dispatch the Event
    FREDispatchStatusEventAsync(self.context, (uint8_t*)INTERSTITIAL_LEFT_APPLICATION, (uint8_t*)[INTER_ID UTF8String]);
}

// =================================================================================================
//	Private Functions
// =================================================================================================

/**
 * Clear the Interstitial
 *
 */
- (void) clearInterstitial
{
    Log(@"clearInterstitial");
    // Clear the Interstitial from memory
    [self.view removeFromSuperview];
    self.view = nil;
    mInterstitialView = nil;
    context = NULL;
    adMobRequest = NULL;
    adMobId = NULL;
    // Clear its Instance
    clearInterstitialInstance();
}

/**
 * Build the Interstitial
 *
 */
- (void) buildInterstitial
{
    Log(@"buildInterstitial");
    
    // Build the Interstitial
    mInterstitialView = [[GADInterstitial alloc] init];
    // Set Banner delegate
    mInterstitialView.delegate = self;
    mInterstitialView.adUnitID = adMobId;
    // Update Load state and start loading the request
    mIsInterstitialLoaded = NO;
    Log(@"isLoaded set to: NO");
    [mInterstitialView loadRequest:adMobRequest];
    Log(@"Creation Completed");
}

// =================================================================================================
//	Public Functions
// =================================================================================================

/**
 * Create the Interstitial
 *
 */
- (void) create
{
    Log(@"Create Interstitial");
    
    // Do not process if the banner was already created
    if(mInterstitialView) return;
    
    Log(@"Creating.....");
    // Initialize the view
    id delegate = [[UIApplication sharedApplication] delegate];
    UIWindow * win = [delegate window];
    [win addSubview:self.view];
    // Disable view touches
    self.view.userInteractionEnabled =NO;
}

/**
 * Remove the Banner
 *
 */
- (void) remove
{
    Log(@"remove");
    // Check if the banner is available
    if(mIsInterstitialLoaded)
    {
        Log(@"removing");
        // Clear the Interstitial from memory
        [self clearInterstitial];
    }
}

/**
 * Cache the Interstitial
 *
 */
- (void) cache
{
    Log(@"Cache Interstitial (unsupported in iOS, Please use create)");
    // Nothing really... Can't cache it
    //mIsInterstitialLoaded = NO;
    //Log(@"isLoaded set to: NO");
    // Set Banner delegate
    //mInterstitialView.delegate = self;
    //[mInterstitialView loadRequest:adMobRequest];
}

/**
 * Check the Interstitial Load State
 *
 */
- (BOOL) isInterstitialLoaded
{
    Log(@"isInterstitialLoaded");
    // return the current load state
    return mIsInterstitialLoaded;
}

/**
 * Show the Interstitial
 *
 */
- (void) show
{
    Log(@"Show Interstitial");

    // Show the Interstitial
    // Enable view Touches
    self.view.userInteractionEnabled =YES;
    // Show the Interstitial
    [mInterstitialView presentFromRootViewController:self];
}

// =================================================================================================
//	Constructor / Deconstructors
// =================================================================================================

/**
 * View Did Load
 * Executed when the view has been loaded in memory
 * Initialize everything which need to be show. (For Example our Interstitial)
 *
 */
- (void)viewDidLoad
{
    Log(@"interstitial viewDidLoad");
    [super viewDidLoad];
    // Build the Banner
    [self buildInterstitial];
}

/**
 * Did Receive Memory Warning
 * Executed when the OS send a memory usage alert.
 * Use it for remove unecessary memory usage.
 *
 */
- (void)didReceiveMemoryWarning
{
    Log(@"interstitial didReceiveMemoryWarning");
    // Releases the view if it doesn't have a superview.
    [super didReceiveMemoryWarning];
}

/**
 * View Did Unload
 * Executed when the view has been unloaded from memory
 *
 */
- (void)viewDidUnload
{
    Log(@"interstitial viewDidUnload");
    [super viewDidUnload];
    mInterstitialView.delegate = nil;
    mInterstitialView = nil;
    context = NULL;
    adMobRequest = NULL;
    adMobId = NULL;
}

/**
 * Dealloc Function
 * Deallocate all the current instances
 *
 */
- (void)dealloc
{
    Log(@"interstitial dealloc");
    [super dealloc];
    /*
    //[self.view removeFromSuperview];
    if(self.view) self.view = nil;
    mInterstitialView = nil;
    context = NULL;
    adMobRequest = NULL;
    adMobId = NULL;
    */
}

@end

File: src/java/src/com/admob/AdMobInterstitial.java
//
//  AdMobAne ANE Extension
//  Android Native Extension
//
//  Copyright (c) 2011-2015 Code lancelot1 Inc. All rights reserved.
//

package com.admob;

//Android Includes
import android.app.Activity;

//Google adMob SDK Includes
import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.InterstitialAd;

/**
 * AdMob Interstitial Class
 * The class will create and manage the Ad Interstitial
 *
 * @author Code lancelot1
 */
public class AdMobInterstitial {
	// Debug Tag
	private static final String CLASS = "AdMobInterstitial - ";

	// Interstitial unit AD Label
	public static final String INTER_ID	= "InterstitialAd";
	
	// Working Instances
	private ExtensionContext mContext;
	private Activity mActivity;
	private InterstitialAd mInterstitialAd;

	// Banner Propriety
    private String mInterstitialId;
	
	/**
	 * Public Getter for Extension Context
	 * 
	 * @return Extension Context Instance
	 **/
	public ExtensionContext getContext() { return mContext; }
	
	/**
	 * Public Getter for Extension Activity
	 * 
	 * @return Extension Activity Instance
	 **/
	public Activity getActivity() { return mActivity; }
	
	/**
	 * AdMob Interstitial Constructor
	 * 
	 * @param ctx Extension context instance
	 * @param act Activity instance
	 * @param interstitialId AdMobId
	 **/
    public AdMobInterstitial(ExtensionContext ctx, Activity act, String interstitialId)
    {
    	// Set Interstitial working Instances
    	mContext		= ctx;
    	mActivity		= act;

    	// Set Interstitial Propriety
    	mInterstitialId	= interstitialId;
    }
    
	/**
	 * Create the interstitial
	 * 
	 */
	public void create() {
    	mContext.log(CLASS+"create interstitial");

		// Create the interstitial.
		mInterstitialAd = new InterstitialAd(mActivity);
		mInterstitialAd.setAdUnitId(mInterstitialId);
		
    	mContext.log(CLASS+"set interstitial Listeners");
		// Add the Listener for the InterstitialAd
		mInterstitialAd.setAdListener(new AdListener() {
			/**
			 * onAdLoaded Callback
			 */
            @Override
            public void onAdLoaded() {
            	// Dispatch event to AS
            	mContext.dispatchStatusEventAsync(ExtensionEvents.INTERSTITIAL_LOADED, INTER_ID);
            }
			/**
			 * onAdFailedToLoad Callback
			 */
            @Override
            public void onAdFailedToLoad(int error) {
            	// Dispatch event to AS
            	mContext.dispatchStatusEventAsync(ExtensionEvents.INTERSTITIAL_FAILED_TO_LOAD, INTER_ID);
            }
			/**
			 * onReceiveAd Callback
			 */
            @Override
            public void onAdOpened() {
            	// Dispatch event to AS
            	mContext.dispatchStatusEventAsync(ExtensionEvents.INTERSTITIAL_AD_OPENED, INTER_ID);
            }
			/**
			 * onReceiveAd Callback
			 */
            @Override
            public void onAdClosed() {
            	// Dispatch event to AS
            	mContext.dispatchStatusEventAsync(ExtensionEvents.INTERSTITIAL_AD_CLOSED, INTER_ID);
            }
			/**
			 * onReceiveAd Callback
			 */
            @Override
            public void onAdLeftApplication() {
            	// Dispatch event to AS
            	mContext.dispatchStatusEventAsync(ExtensionEvents.INTERSTITIAL_LEFT_APPLICATION, INTER_ID);
            }
        });
		
    	mContext.log(CLASS+"set interstitial Request");
		// Get the interstitial Request for adMob
		AdRequest request = mContext.getRequest();
		//AdRequest request = getRequest();
		// Load the ad
    	mContext.log(CLASS+"load interstitial Request");
		mInterstitialAd.loadAd(request);
	}
	
	/**
	 * Remove the Interstitial
	 * 
	 */
	public void remove() {
    	mContext.log(CLASS+"remove");
    	mContext		= null;
    	mActivity		= null;
    	mInterstitialId	= null;
	}
	
	/**
	 * Cache the interstitial
	 * 
	 */
	public void cache() {
    	mContext.log(CLASS+"cache interstitial");
    	mContext.log(CLASS+"set interstitial Request");
		// Get the interstitial Request for adMob
		AdRequest request = mContext.getRequest();
		//AdRequest request = getRequest();
		// Load the ad
    	mContext.log(CLASS+"load interstitial Request");
		mInterstitialAd.loadAd(request);
	}
	
	/**
	 * Check if the Interstitial is loaded
	 * 
	 * @return Interstitial Load state
	 */
	public Boolean isLoaded() {
    	mContext.log(CLASS+"isLoaded: "+mInterstitialAd.isLoaded());
		// Check the Interstitial
		return mInterstitialAd.isLoaded();
	}
	
	/**
	 * Show the interstitial
	 * 
	 */
	public void show() {
    	mContext.log(CLASS+"show");
		// Show the Interstitial
		mInterstitialAd.show();
	}
}
